Migrate to tiff crate

gschulze opened the pull request:
This PR migrates everything to the `tiff` crate, doing significant simplification and cleanup along the way.

- The `TIFF` struct has been renamed to `GeoTIFF`. The functionality in `reader.rs` and `lowlevel.rs` has been removed; reading TIFF data is accomplished through the `read` function in the struct implementation, which directly calls the functionality provided by the `tiff` crate.
- Obtaining sample values is provided by the `get_value_at<T>(x: usize, y: usize, sample: usize)` function. The desired output type is specified by the generic parameter `T`. The `sample` parameter has been added to the function signature to allow retrieving data from multi-channel TIFFs. The function panics if the value is not representable as the desired output type or the requested sample is out of range.

This refactoring aims to reduce the code to the bare minimum needed to provide the functionality covered by the integration tests. The next step is to apply the correct coordinate transformation defined by the corresponding GeoTIFF tags.

kylebarron commented:
> This refactoring aims to reduce the code to the bare minimum needed to provide the functionality covered by the integration tests. The next step is to apply the correct coordinate transformation defined by the corresponding GeoTIFF tags.

In case you want a reference for parsing the geokey tags, I have an implementation here: https://github.com/developmentseed/aiocogeo-rs/blob/main/src/geo_key_directory.rs (it doesn't use the `tiff` crate)

weiji14 commented on src/lib.rs:
```diff
@@ -1,40 +1,97 @@
+/// The image data has a size of width * length * num_samples
+#[derive(Debug)]
+pub struct GeoTIFF {
+    pub dimensions: (usize, usize),
```
Thoughts on separating this into `width` and `height`? I always get tripped up on whether it's x/y or y/x.

gschulze replied:
I can do that, of course. I just used the format of the underlying `tiff` crate. We should maybe also make it clear that we are referring to raster coordinates as opposed to model coordinates, calling it  `rasterWidth` and `rasterHeight`.

weiji14 replied:
Yep, `rasterWidth` and `rasterHeight` sounds good to me!

weiji14 commented on src/lib.rs:
```diff
@@ -1,40 +1,97 @@
+#[derive(Debug)]
+pub struct GeoTIFF {
+    pub dimensions: (usize, usize),
+    pub num_samples: usize,
```
I know you've gone back and forth on this, but just want to check if we're ok with settling on `num_samples` and not `num_bands`?

gschulze replied:
Yeah, I wasn't sure what is the most appropriate terminology to use here
- On the `TIFF` file format level, they refer to `samples`.
- In their description of the raster data model, [GDAL](https://gdal.org/en/latest/user/raster_data_model.html) calls it `bands`.
- Image manipulation software usually calls it `channels.`
- `layer` is another possible term mentioned by GDAL.

The reason why I chose `sample` over `band` was because the origin of `band` comes from the bandwidth of the color spectrum that the sample is supposed to cover, which does not apply to all scenarios where raster datasets are used. I have no absolute preference here; we can also change the name later if we discover that it does not fit.

weiji14 replied:
Yeah, I'm torn as well, the remote sensing part of me wants to call it `bands`, but the computer vision part of me likes `channels`. Let's stick with `samples` then following TIFF, though it feels very statistical for some reason :laughing:

weiji14 commented on src/lib.rs:
```diff
@@ -1,40 +1,97 @@
+mod image_data;
 
-/// The GeoTIFF library reads `.tiff` files.
+macro_rules! assert_type {
+    ($result: expr, $actual: ty, $expected: ty) => {
```
This macro is :fire:! Could we rename it from `assert_type` to something else though? I realize it does do a check/assertion, but it's purpose is more about converting `ImageData` to a primitive type, so maybe `to_primitive_type` or something like that?

gschulze replied:
I agree; I was a bit lazy on this one. Since the macro is not doing the conversion but just unpacking the result if the type is correct, would maybe something like `unwrap_primitive_type` be more appropriate?

weiji14 replied:
Yep, we can go with `unwrap_primitive_type`, though we might remove the `unwrap` call at some point :wink:

weiji14 commented on src/lib.rs:
```diff
@@ -1,40 +1,97 @@
+            ..
+        } = self;
+
+        if sample >= *num_samples {
+            panic!(
+                "sample out of bounds: the number of samples is {} but the sample is {}",
+                num_samples, sample
+            )
+        }
```
Could we raise an error (e.g. [`TiffError::LimitsExceeded`](https://docs.rs/tiff/0.9.1/tiff/enum.TiffError.html#variant.LimitsExceeded)) here instead of panicking?

gschulze replied:
I thought this was similar to accessing an array element, which also panics if the index is out of bounds. Also, returning a result might make it more cumbersome to use as a library consumer. What should be checked in addition is whether `x` and `y` are less than `width` and `height`, respectively. However, this will change again if we integrate coordinate transformations.

If we decide on returning a result from this function, I'd vote for having a separate `GeoTiffError` type. As far as I understand, the `TiffError` deals with errors related to encoding/decoding, whereas here, we are handling a potential programming error.

weiji14 replied:
> I thought this was similar to accessing an array element, which also panics if the index is out of bounds.

A panic can happen if you access using `data[index]`, but not if you use `data.get(index)` which returns an `Option` (see https://doc.rust-lang.org/std/vec/struct.Vec.html#indexing and https://doc.rust-lang.org/std/primitive.slice.html#method.get.

> What should be checked in addition is whether `x` and `y` are less than `width` and `height`, respectively. However, this will change again if we integrate coordinate transformations.

Yes, ok with leaving `x > width` and `y > height` unchecked for now, and do it after coordinate transforms are implemented.

> If we decide on returning a result from this function, I'd vote for having a separate `GeoTiffError` type. As far as I understand, the `TiffError` deals with errors related to encoding/decoding, whereas here, we are handling a potential programming error.

Don't want us to get sidetracked with implementing `GeoTIFFError` and `GeoTIFFResult` (though I'm sure that will come at some point). For now, I'm happy to compromise on keeping the panic here since it matches previous behaviour, and we can refactor later to return a `Result` instead of panic.

gschulze replied:
Just got distracted by the looks of `GeoTIFFError` and `GeoTIFFResult`. I think it should be written `GeoTiffError` and `GeoTiffResult`, according to the Rust naming conventions. Consequently, I believe the type in the crate should also be called `GeoTiff`.

weiji14 reviewed (commented):
Thanks @gschulze for pulling this together! Just some minor comments for now, mostly related to variable names (yes, naming is hard), and a `panic!` that could probably be removed.

I'll be doing a Rust pair programming session with a colleague later, and might try to convince him to go though this PR so there might be a few more comments coming!

weiji14 commented on src/lib.rs:
```diff
@@ -1,40 +1,97 @@
+            ..
+        } = self;
+
+        if sample >= *num_samples {
```
Could also match by turning `sample` into a reference instead of dereferencing `num_samples`?

```suggestion
        if &sample >= num_samples {
```

gschulze replied:
Will do!

weiji14 commented on src/lib.rs:
```diff
@@ -1,40 +1,97 @@
+                    type_name::<$expected>()
+                )
+            })
+            .unwrap()
```
The previous `get_value_at` method was returning a value directly, so using `.unwrap()` here makes sense to match the same behaviour. But maybe it would be better to return a `TiffResult` once the panic at https://github.com/georust/geotiff/pull/17/files#r1746172003 is removed.

gschulze replied:
As mentioned above, I'd use a separate `GeoTiffError` type for errors specific to this crate.

weiji14 replied:
Ok with leaving this `.unwrap()` for now as mentioned at https://github.com/georust/geotiff/pull/17#discussion_r1746581439.

weiji14 commented on src/lib.rs:
```diff
@@ -1,40 +1,97 @@
-    }
+/// The image data has a size of width * length * num_samples
+#[derive(Debug)]
+pub struct GeoTIFF {
+    pub dimensions: (usize, usize),
+    pub num_samples: usize,
+    image_data: ImageData,
+}
```
Another thought I had was whether we should just expose the tiff [`Decoder`](https://docs.rs/tiff/0.9.1/tiff/decoder/struct.Decoder.html), and have the attributes like `dimensions`, `num_channels`, etc be accessible via methods only. Thoughts on this?

gschulze replied:
Given that `Decoder<R>` consumes an underlying stream of type `R`, I think this would prevent us from implementing write support in the future.

For instance, I have a use case where I need to create a new, empty GeoTIFF, which is then populated with data. Finally, I want to save it to a file.

weiji14 replied:
Ah yes, I've been thinking in read-only mode for now, but you're correct that we can't put in `Decoder<R>` if implementing write support later.

> For instance, I have a use case where I need to create a new, empty GeoTIFF, which is then populated with data. Finally, I want to save it to a file.

Secretly, I would love to see this implemented soon, mostly just so we can write some roundtrip read/write tests and not have to store sample GeoTIFF files in this repo!

gschulze commented:
Hi @weiji14, I've incorporated your feedback and force-pushed the branch. I've renamed the `GeoTIFF`type to `GeoTiff`, to follow the Rust naming conventions, if you don't mind.

I wonder whether we should rename `ImageData` to `RasterData` to align more closely with `raster_width` and `raster_height`. What do you think?

Besides that, I think the PR is ready for merging.

weiji14 reviewed (approved):
> Hi @weiji14, I've incorporated your feedback and force-pushed the branch. I've renamed the `GeoTIFF`type to `GeoTiff`, to follow the Rust naming conventions, if you don't mind.

Awesome, thanks @gschulze! I do need to check https://rust-lang.github.io/api-guidelines/naming.html more often :smile: Will merge this in after you do the `ImageData` -> `RasterData` rename (probably after the weekend).

On the force-pushing, I'm a little against it as a reviewer because it undos all the `Viewed` checkboxes I tick in the 'Files Changed' tab, and somewhat forces me to have to carefully look over everything again if I'm checking things locally on my diff viewer or on GitHub mobile (it's a little better on the GitHub webpage where I can click on 'Compare'). Also, I tend to squash merge so many little commits don't matter, but if you prefer regular merges or rebase+merge, I can go with that since you've only got one commit here.

gschulze commented:
> On the force-pushing, I'm a little against it as a reviewer because it undos all the `Viewed` checkboxes I tick in the 'Files Changed' tab, and somewhat forces me to have to carefully look over everything again if I'm checking things locally on my diff viewer or on GitHub mobile (it's a little better on the GitHub webpage where I can click on 'Compare'). Also, I tend to squash merge so many little commits don't matter, but if you prefer regular merges or rebase+merge, I can go with that since you've only got one commit here.

Understandable. Just tell me the workflow you prefer. I usually try to do commits that are somehow semantically complete. That's why I try to avoid "Typo"-level commits. I'm okay with you squashing them together later.

gschulze commented:
@weiji14 Renamed 'ImageData' to 'RasterData'
